### PR TITLE
Ports VG's DeityLink's new explosion method

### PR DESCRIPTION
**Pull request by RemieRichards**
- Ports /vg/ DeityLink's new explosion effect
- Blast doors are more resistant to explosions
- Engine floors (black hexagon ones) are more resistant to explosions
- Explosions can now be affected by Walls and Doors (Cheaper than recursive explosions!)
- Adds config option for new explosion effect (Defaults to on, It's nicer, it's still cheap to run, etc.)
- Adds debug verb for displaying the range and effect of various explosions

Feature? uhh I don't know what you're talking about, this fixes <i>realism.</i> :P 

I even took the gif, Muahahaha:

@DeityLink

### CODE REVIEW

**Comment by Chocobro** [comment]
this fixes the sprite bug that our explosion effects kind of sucked

**Comment by PKPenguin321** [comment]
looks seriously awesome (but probably not appropriate for a freeze :snowflake: :cry: )

**Comment by d3athrow** [comment]
i ban u

**Comment by RemieRichards** [comment]
@d3athrow BKAWWWWK

**Comment by DeityLink** [comment]
>porting my features before we even merge them on vg

I, I guess I'm flattered.

**Comment by RemieRichards** [comment]
@DeityLink if it's good, I take it, so yes, flattery very much implied.

**Comment by GunHog** [comment]
When I saw this PR on the other codebase, I knew RR would bring it here :+1:

**Comment by Supermichael777** [comment]
rr dismeberment when? lol.

**Comment by Anonus** [comment]
What are the major differences between this and erro's weird recursive explosions that we never use?

**Comment by RemieRichards** [comment]
@Anonus it's pretty much the same effect, except this doesn't lag to shit (I didn't notice ANY performance impacts at all, but that was just "by eye" testing)

**Comment by Anonus** [comment]
"Faster" is good enough for me.

**Comment by DeityLink** [comment]
@Anonus https://github.com/d3athrow/vgstation13/pull/4347#issuecomment-100296433

**Comment by Incoming5643** [comment]
The implicit nerf to bomb radius should be shored up with a buff to the bombcap

:bomb: :bomb: :bomb: :microscope: :bomb: :bomb: :bomb:

**Comment by Lumipharon** [comment]
For once I actually agree that increasing the cap could actually make sense.

**Comment by DeityLink** [comment]
On vg Pomf increased the cap to dev 7.

**Comment by Lumipharon** [comment]
Also for better or worse, chem/slime nades would be a lot weaker/harder to make cap, where as toxin bombs it's pretty easy to make very big booms.

**Comment by Jordie0608** [comment]
Does this hook in with #9303?

**Comment by Cheridan** [comment]
>looks seriously awesome (but probably not appropriate for a freeze :snowflake: :cry: )

If you wanted an excuse to merge this you could just call it an optimization to recursive explosions.

**Comment by DeityLink** [comment]
More like a replacement, I didn't touch explosion_recursive.dm at all.

**Comment by Anonus** [comment]
Just claim it's an optimization so we can claim it's a fix and not a feature

**Comment by daman997** [comment]
i can dig it :+1:

**Comment by optimumtact** [comment]
performance impact?

**Comment by RemieRichards** [comment]
@optimumtact 
> it's pretty much the same effect, except this doesn't lag to shit (I didn't notice ANY performance impacts at all, but that was just "by eye" testing)

**Comment by optimumtact** [comment]
Oops, Didn't notice :relaxed: :+1: from me

File: code/game/machinery/doors/poddoor.dm [hunk 1]
```diff
@@ -86,318 +87,27 @@
 	operating = 0
 
 
+//"BLAST" doors are obviously stronger than regular doors when it comes to BLASTS.
```
Comment (GunHog) [inline code/game/machinery/doors/poddoor.dm]: https://www.youtube.com/watch?v=huDHb45P7v0

**Comment by MrStonedOne** [comment]
This actually lowers lag: https://github.com/ParadiseSS13/Paradise/pull/979#issuecomment-100555928

Less items to delete because of walls blocking explosions. Almost all explosions happen near walls, a good chunk near r-walls.

And this isn't quite what recursive explosion does, because recursive explosions allowed you to "shape" a charge. Meaning that if you put a bomb with rwalls all around it but one direction, it explodes out further in that one direction as well as getting blocked in the other directions.

Still cool none the less.

File: code/game/machinery/doors/poddoor.dm (hunk 1 above)
Comment (DeityLink) [inline code/game/machinery/doors/poddoor.dm]: hah, I didn't remember that scene when I wrote that comment.

**Comment by ZomgPonies** [comment]
As @MrStonedOne said, it does lower the total explosive potential of a charge as walls absorb some of the blast rather than reflecting it, but it's still quite a step up in terms of behavior compared to non-recursive explosions.

**Comment by Incoming5643** [comment]
runtime error: Cannot read null.x
proc name: Check Bomb Impact (/client/proc/check_bomb_impacts)
  source file: explosion.dm,188
  usr: Imani White (/mob/living/carbon/human)
  src: Incoming (/client)
  call stack:
Incoming (/client): Check Bomb Impact()

Thrown at me whenever I tried to use Check Bomb Impact.

Additionally it doesn't seem to actually work right now, all bombs still process in the old way even with the config on.

**Comment by ZomgPonies** [comment]
Well shit. Look at all the people who ported it too :cry:

**Comment by DeityLink** [comment]
That's what happens when you port a port :P (There are no issues with either the debug command of the explosions themselves on vg)

**Comment by Miauw62** [comment]
That's what you get for taking unfinished features asap
Op 13-mei-2015 03:07 schreef "ZomgPonies" <notifications@github.com>:

**Comment by DeityLink** [comment]
How is it unfinished?

**Comment by Miauw62** [comment]
Well, a buggy port. I was replying to zomgponies
Op 13-mei-2015 10:02 schreef "DeityLink" <notifications@github.com>:

**Comment by RemieRichards** [comment]
Uh @Incoming5643 I was able to use Check Bomb Impact just fine, 0 errors.
I was able to use bombs that did indeed react to their environment.

I have 0 clue how it's broken, because I tested this crap extensively.

@DeityLink I ported straight from /vg/, so if yours is perfect the issue is on our end.

**Comment by ZomgPonies** [comment]
@RemieRichards We have the same runtimes on Paradise and I'm afraid the behavior issues are probably the same.

**Comment by Incoming5643** [comment]
Are there any new .dm files you may have forgotten to commit because they were previously untracked in your local copy?

**Comment by RemieRichards** [comment]
@Incoming5643 No, no new files.
If I'd have forgotten a file there'd be an empty .dme here.

File: code/game/objects/explosion.dm [hunk 2]
```diff
@@ -138,3 +150,74 @@ proc/explosion(turf/epicenter, devastation_range, heavy_impact_range, light_impa
...
 		tile.ex_act(2)
+
+
+/client/proc/check_bomb_impacts()
+	set name = "Check Bomb Impact"
+	set category = "Debug"
+
+	var/newmode = alert("Use reactionary explosions?","Check Bomb Impact", "Yes", "No")
+
+	var/turf/epicenter = get_turf(src)
```
Comment (JJRcop) [inline code/game/objects/explosion.dm]: Source of Incoming's runtime.

Comment (Incoming5643) [inline code/game/objects/explosion.dm]: Can be resolved by making src usr instead (I know usr is bad but it's a debug verb I think we can get away with it)

**Comment by Incoming5643** [comment]
I might be wrong on the bombs not respecting the new mechanics, you can probably disregard that point, it's just the debug verb that was borked.

File: code/game/objects/explosion.dm (hunk 2 above)
Comment (optimumtact) [inline code/game/objects/explosion.dm]: >usr
REEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEeee

**Comment by RemieRichards** [comment]
@Incoming5643 I'll do some more tests tonight.

File: code/game/objects/explosion.dm (hunk 2 above)
Comment (RemieRichards) [inline code/game/objects/explosion.dm]: Can probably just find the turf of client.mob instead.

**Comment by RemieRichards** [comment]
@tgstation/commit-access @Incoming5643 was indeed wrong, Bombs do respect the new mechanics
I have also now fixed the debug verb runtime.

**Comment by Incoming5643** [comment]
glad I was right about being wrong

**Comment by RemieRichards** [comment]
@Incoming5643 :P

**Comment by RemieRichards** [comment]
@ZomgPonies
@Crystalwarrior
@HoaxIT

Just to update you lot, this (from my further testing) has no behaviour issues, and was infact just a debug verb runtime.

**Comment by DeityLink** [comment]
Wasn't the runtime maptext var related anyway? I told you that colors were enough! :(

**Comment by RemieRichards** [comment]
@DeityLink nah, it was with finding epicentre (granted the runtime Incoming got I -never- got, but the "fix" I applied should avoid the issue)

**Comment by Crystalwarrior** [comment]
Yeah no problem, I discovered the runtime before meeting the explosion method.

**Comment by RemieRichards** [comment]
@tgstation/commit-access mergy mergy pleaserino

**Comment by DeityLink** [comment]
Honk

**Comment by HoaxIT** [comment]
Honk

Sent from a not-so-smartphone. Anything written herein that you find misspelled, objectionable, incoherent, dim-witted, plagiarized or legally actionable should be attributed to the phone manufacturer, Chinese hackers, or PRISM.

**Comment by Anonus** [comment]
So, the bombcap got raised to compensate for this except this isn't merged yet, so the bombcap is 5,10,20 and not falling off with more stuff in the way, which is probably a bad thing.

**Comment by Lumipharon** [comment]
I saw bombs today and admins told me we still have the old cap, so that's not an issue yet, I think.

This probably should be merged sooner rather then later though.

**Comment by Razharas** [comment]
@RemieRichards 
@Cheridan 
This needs balance decision and conflict fix

**Comment by RemieRichards** [comment]
@Razharas "Controversial" my ass, This has been discussed for weeks now, it was only not merged because of the Freeze (and the conflict I'm fixing atm)

**Comment by DeityLink** [comment]
Wait you guys still haven't merged it?

**Comment by RemieRichards** [comment]
@DeityLink I opened it during a feature freeze so I wouldn't forget to port it, so not yet lol.

**Comment by Lumipharon** [comment]
Please merge and buff the bomb cap a bit. This is neat, and makes explosions way more realistic.

**Comment by HoaxIT** [comment]
We've been running this on HippieStation for quite some time, i'd suggest you up the bomb cap when merging this.

**Pull request merged**